### ISSUE TITLE
Dict in njitted function

### ISSUE THREAD

**Issue by BEpresent** [numba/numba#6488]
Minimum example:

```
d_test = Dict.empty(
    key_type=types.float64,
    value_type=types.float64,
)

d_test['a'] = 0.1
d_test['b'] = 0.2


@nb.njit()
def nb_dict_test():
    v1 = d_test['a']
    v2 = d_test['b']
    summed_v = v1 + v2
    return summed_v

nb_dict_test()
```

or 

```
d_test = Dict.empty(
    key_type=types.float64,
    value_type=types.float64,
)

d_test[0.1] = 0.1
d_test[0.2] = 0.2


@nb.njit()
def nb_dict_test():
    v1 = d_test[0.1]
    v2 = d_test[0.2]
    summed_v = v1 + v2
    return summed_v

nb_dict_test()
```

do not work with  error message
```
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: ensure IR is legal prior to lowering)
The use of a DictType[unicode_type,float64]<iv=None> type, assigned to variable 'd_test' in globals, is not supported as globals are considered compile-time constants and there is no known way to compile a DictType[unicode_type,float64]<iv=None> type as a constant.
File "<string>", line 12:
<source missing, REPL/exec in use?>
```

Not sure if this is indeed not implemented or there is an error in my implementation?

**Comment by gmarkall** [comment]
It is a deviation from Python semantics in Numba that globals are treated as constants: https://numba.readthedocs.io/en/stable/reference/pysemantics.html?highlight=global#global-and-closure-variables - however, I think it ought to be possible to support your use cases, since they don't mutate the `Dict` inside the jitted function.

I'll labelling this as a feature request, to support read-only access to global typed dicts in nopython mode.

**Comment by stuartarchibald** [comment]
> Putting the dict into the function does work, but this is not the desired solution as it is slow for my case.

In what way is this slow? Is it that a dictionary has to be built?

**Comment by BEpresent** [comment]
Indeed, the above is just a toy example. The creation of my dictionary itself takes several minutes and my intention is to build it once and then access it with a different @nb.njit() function multiple times.

**Comment by stuartarchibald** [comment]
> Indeed, the above is just a toy example. The creation of my dictionary itself takes several minutes and my intention is to build it once and then access it with a different @nb.njit() function multiple times.

Is it built dynamically or is it the same dictionary every time?

**Comment by BEpresent** [comment]
It is the same dictionary, so it should be considered a constant. This is why I would assume that this should be possible.

**Comment by stuartarchibald** [comment]
> It is the same dictionary, so it should be considered a constant. This is why I would assume that this should be possible.

Can you just define it inline, like:
```python
def my_table():
    d = {1:2, 3:4}
```
?

**Comment by BEpresent** [comment]
It is very large with over 100k entries, but I guess I could try with a dict comprehension ? I could test this. Would this make numba accept it ?

**Comment by stuartarchibald** [comment]
What matters is where the dictionary is built, there's a trade off between making the compiler work hard to build the dictionary in a function and then caching that function, and building the dictionary in the interpreter and then passing it around. At least if it's defined in a function it can be reused relatively easily. What's the data like to generate the dictionary? Is it an already existing python dictionary or can the way to build it be described programmatically?

**Comment by BEpresent** [comment]
The dictionary can be built programmatically based on Numpy arrays containing some vectors. It involves some number-crunching and the goal is to have some kind of lookup table to/dictionary to just look up the result (just a float, the key can be anything). I do not even need to be a dictionary, just some method for memoization to avoid recalculation of the answers (here https://stackoverflow.com/questions/63107727/numba-compatable-memoization someone else had a similar problem). The only constraint is that the lookup needs to be inside a function that can be 'njitted' as this is required by the library I use.

**Comment by pengchaoandy** [comment]
> It is a deviation from Python semantics in Numba that globals are treated as constants: https://numba.readthedocs.io/en/stable/reference/pysemantics.html?highlight=global#global-and-closure-variables - however, I think it ought to be possible to support your use cases, since they don't mutate the `Dict` inside the jitted function.
> 
> I'll labelling this as a feature request, to support read-only access to global typed dicts in nopython mode.

If there is a workaround to support this feature in numba 0.52?
I tried to hack lower_assign by return llvmlite.ir.values.Constant if value.value is a typed.Dict, but not working.

**Comment by legendagain** [comment]
Hello! Just seconding this feature request and checking in if you have found a workaround to this? I'm currently facing the same issue (using a fixed dictionary external to njitted function) and this feature would be extremely useful.
